## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [19]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [43]:
# View the column names in the data
column_names = df.columns
column_names

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [45]:
# Use the describe function to gather some basic statistics
description = df.describe()
description

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [69]:
# Use this space to do any additional research
# and familiarize yourself with the data.
data_types = df.dtypes
non_null_counts = df.count()

data_info = pd.DataFrame({
    '#': range(1, len(column_names) + 1),
    'Data Type': data_types,
    'Non-Null Count': non_null_counts
})

data_info


# Data Type  Non-Null Count
first         1    object           54639
last          2    object           54639
job           3    object           54639
phone         4    object           54639
email         5    object           54639
client_id     6     int64           54639
order_id      7     int64           54639
order_date    8    object           54639
order_week    9     int64           54639
order_year   10     int64           54639
item_id      11    object           54639
category     12    object           54639
subcategory  13    object           54639
unit_price   14   float64           54639
unit_cost    15   float64           54639
unit_weight  16   float64           54639
qty          17     int64           54639
line_number  18     int64           54639

In [71]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
top_three_categories = category_counts.head(3)
top_three_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [105]:
# For the category with the most entries, which subcategory had the most entries?
top_three_df = df[df['category'].isin(top_three_categories)]
subcategory_counts = top_three_df['subcategory'].value_counts()
top_subcategory = subcategory_counts.idxmax()
top_subcategory

'bathroom supplies'

In [99]:
# Which five clients had the most entries in the data?
client_counts = df['client_id'].value_counts()
client_counts.head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [109]:
# Store the client ids of those top 5 clients in a list.
top_five_clients = client_counts.head(5)
top_five_client_ids = top_five_clients.index.tolist()
top_five_client_ids

[33615, 66037, 46820, 38378, 24741]

In [113]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()
top_client_df = df[df['client_id'] == top_client]
total_units_ordered = top_client_df['qty'].sum()
total_units_ordered

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [135]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['qty', 'unit_price', 'line_subtotal']].head()

qty  unit_price  line_subtotal
0  105     1096.80      115164.00
1   21       24.95         523.95
2   39       13.52         527.28
3   29       36.42        1056.18
4   20      195.10        3902.00

In [137]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
df[['order_id', 'unit_weight', 'shipping_price']].head()


order_id  unit_weight  shipping_price
0   8953482         7.50            75.0
1   8069089         1.49            14.9
2   1902144         1.68            16.8
3   9031802         1.23            12.3
4   1322274        46.43           464.3

In [139]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df['sales_tax'] = 0.0925
df['total_price'] = df['line_subtotal'] + df['shipping_price']
df['total_price_with_sales_tax'] = df['total_price'] * (1 + df['sales_tax'])

df[['qty', 'unit_price', 'line_subtotal', 'unit_weight', 'shipping_price', 'total_price_with_sales_tax']].head()

qty  unit_price  line_subtotal  unit_weight  shipping_price  \
0  105     1096.80      115164.00         7.50            75.0   
1   21       24.95         523.95         1.49            14.9   
2   39       13.52         527.28         1.68            16.8   
3   29       36.42        1056.18         1.23            12.3   
4   20      195.10        3902.00        46.43           464.3   

   total_price_with_sales_tax  
0               125898.607500  
1                  588.693625  
2                  594.407400  
3                 1167.314400  
4                 4770.182750

In [133]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['cost_each_line'] = df['unit_price'] * df['qty'] + df['shipping_price']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  shipping_price  sales_tax  total_price  \
0      115164.00            75.0     0.0925    115239.00   
1         523.95            14.9     0.0925       538.85   
2         527.28            16.8     0.0925       544.08   
3        1056.18            12.3     0.0925      1068.48   
4        3902.00           464.3     0.0925      4366.30   

   total_price_with_sales_tax  cost_each_line  
0               125898.607500       115239.00  
1                  588.693625          538.85  
2                  594.407400          544.08  
3                 1167.314400         1068.48  
4                 4770.182750         4366.30  

[5 rows x 24 columns]

In [141]:
# Create a column for the profit of each line using line cost and line price

df['profit'] = df['total_price_with_sales_tax'] - df['cost_each_line']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  line_subtotal  \
0          17        2023  ...        7.50  105           1      115164.00   
1          20        2023  ...        1.49   21           0         523.95   
2           4        2023  ...        1.68   39           6         527.28   
3          17        2023  ...        1.23   29           3        1056.18   
4          21        2023  ...       46.43   20           1        3902.00   

   shipping_price  sales_tax  total_price  total_price_with_sales_tax  \
0            75.0     0.0925    115239.00               125898.607500   
1            14.9     0.0925       538.85                  588.693625   
2            16.8     0.0925       544.08                  594.407400   
3            12.3     0.0925      1068.48                 1167.314400   
4           464.3     0.0925      4366.30                 4770.182750   

   cost_each_line        profit  
0       115239.00  10659.607500  
1          538.85     49.843625  
2          544.08     50.327400  
3         1068.48     98.834400  
4         4366.30    403.882750  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [185]:
# Check your work using the totals above

order_ids = [2742071, 2173913, 6128929]

for order_id in order_ids:
    try:
        total_price = df.loc[df['order_id'] == order_id, 'total_price'].values[0]
        print(f"Order ID: {order_id}, Total Price: {total_price}")
    except IndexError:
        print(f"Order ID {order_id} not found in the DataFrame.")
    

Order ID: 2742071, Total Price: 535.03
Order ID: 2173913, Total Price: 7745.5
Order ID: 6128929, Total Price: 8966.85


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [189]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
total_spending = df.groupby('client_id')['cost_each_line'].sum()
top_clients_by_spending = total_spending.nlargest(5)
top_clients_by_spending


client_id
78965    5.950544e+08
85931    2.071955e+08
38473    1.636283e+08
86721    1.015892e+08
84344    9.287877e+07
Name: cost_each_line, dtype: float64

In [213]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

summary_dfs = []
top_clients_ids = [78965, 85931, 38473, 86721, 84344]
for client_id in top_clients_ids:

    client_data = df[df['client_id'] == client_id]
    total_units_purchased = client_data['qty'].sum()
    total_shipping_price = client_data['shipping_price'].sum()
    total_revenue = client_data['total_price_with_sales_tax'].sum()
    total_profit = client_data['profit'].sum()
    
    summary_dfs.append({
        'Client': client_id,
        'Total Units Purchased': total_units_purchased,
        'Total Shipping Price': total_shipping_price,
        'Total Revenue': total_revenue,
        'Total Profit': total_profit
    })

summary_df = pd.DataFrame(summary_dfs)

summary_df

Client  Total Units Purchased  Total Shipping Price  Total Revenue  \
0   78965                4033302                7383.3   6.500970e+08   
1   85931                 644423                2763.0   2.263611e+08   
2   38473                 402770                7832.2   1.787640e+08   
3   86721                1010426                6154.6   1.109862e+08   
4   84344                 466968                6944.0   1.014701e+08   

   Total Profit  
0  5.504253e+07  
1  1.916559e+07  
2  1.513562e+07  
3  9.397004e+06  
4  8.591286e+06

In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995